core/commands/repo.go

```diff
@@ -406,3 +413,167 @@ var repoVersionCmd = &oldcmds.Command{
+	SizeTotal int `json:",omitempty"`
+}
+
+var repoHasCmd = &cmds.Command{
+	Helptext: cmdkit.HelpText{
+		Tagline:          "Show if an object is available locally",
+		ShortDescription: ``,
+	},
+	Arguments: []cmdkit.Argument{
+		cmdkit.StringArg("key", true, true, "Key(s) to check for locality.").EnableStdin(),
+	},
+	Options: []cmdkit.Option{
+		cmdkit.BoolOption("recursive", "r", "Check recursively the graph of objects to build more precise stats"),
+	},
+	Run: func(req cmds.Request, res cmds.ResponseEmitter) {
+		n, err := req.InvocContext().GetNode()
+		if err != nil {
+			res.SetError(err, cmdkit.ErrNormal)
+			return
+		}
+
+		recursive, _, err := req.Option("recursive").Bool()
+		if err != nil {
+			res.SetError(err, cmdkit.ErrNormal)
+			return
+		}
+
+		// Decode all the keys
+		var keys []*cid.Cid
+		for _, arg := range req.Arguments() {
+			c, err := cid.Decode(arg)
+			if err != nil {
+				res.SetError(err, cmdkit.ErrNormal)
+				return
+			}
+
+			keys = append(keys, c)
+		}
+
+		// an offline DAGService will not fetch from the network
+		offlineDag := dag.NewDAGService(bservice.New(
+			n.Blockstore,
+			offline.Exchange(n.Blockstore),
+		))
+
+		ctx := n.Context()
+
+		for _, k := range keys {
+			root, err := offlineDag.Get(ctx, k)
+			if err != nil && err != dag.ErrNotFound {
+				res.SetError(err, cmdkit.ErrNormal)
+				return
+			}
+
+			hasRoot := err != dag.ErrNotFound
+
+			if !hasRoot || !recursive {
+				// we don't have precise size information
+				res.Emit(&localityOutput{Hash: k.String(), Local: hasRoot})
+				continue
+			}
+
+			local, sizeLocal, sizeTotal, err := walkBlock(ctx, offlineDag, root)
+			if err != nil {
+				res.SetError(err, cmdkit.ErrNormal)
+				return
+			}
+
+			res.Emit(&localityOutput{
+				Hash:      k.String(),
+				Local:     local,
+				SizeLocal: sizeLocal,
+				SizeTotal: sizeTotal,
+			})
+		}
+	},
+	Encoders: cmds.EncoderMap{
+		cmds.Text: cmds.MakeEncoder(func(req cmds.Request, w io.Writer, v interface{}) error {
+			lo, ok := v.(*localityOutput)
+			if !ok {
+				return e.TypeErr(lo, v)
+			}
+
+			recursive, _, err := req.Option("recursive").Bool()
+			if err != nil {
+				return err
+			}
+
+			var state string
+			if recursive {
+				state = "complete"
+				if !lo.Local {
+					state = "incomplete"
+				}
+				if lo.SizeTotal <= 0 {
+					state = "unknow"
+				}
+			} else {
+				if lo.Local {
+					state = "local"
+				} else {
+					state = "non local"
+				}
+			}
+
+			if lo.SizeTotal > 0 {
+				fmt.Fprintf(w, "%s\t%s\t%s of %s (%.2f%%)\n",
+					lo.Hash,
+					state,
+					humanize.Bytes(uint64(lo.SizeLocal)),
+					humanize.Bytes(uint64(lo.SizeTotal)),
+					100.0*float64(lo.SizeLocal)/float64(lo.SizeTotal),
+				)
+			} else {
+				fmt.Fprintf(w, "%s\t%s\n", lo.Hash, state)
+			}
+
+			return nil
+		}),
+	},
+	Type: localityOutput{},
+}
+
+func walkBlock(ctx context.Context, dagserv dag.DAGService, merkleNode node.Node) (bool, int, int, error) {
```

Comment:
I'd just call `merkleNode` a `nd` as that's what it's usually called

Reply:
This should also probably get `recursive` or `depth` pram as it looks like that we traverse the tree even if --recursive is not specified

Reply:
No, this function is not called without `recursive`